### PR TITLE
`Programming exercises`: Automatically renew personal access token for GitLab

### PR DESCRIPTION
**Pull request by R3dError**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **Gitlab** and **Jenkins**.

### Motivation and Context
GitLab recently introduced a lifetime limit for personal access tokens, see https://about.gitlab.com/blog/2023/10/25/access-token-lifetime-limits/. All previously created personal access tokens will expire on May 14, 2024. Therefore, personal access tokens now have to be periodically renewed, as described by #7954.

### Description
As it has been decided in #7954, we now store the expiration date of VCS access tokens in the user table. The class `VcsTokenRenewalService` implements the periodic renewal of expiring or already expired access tokens. Using the abstract class `VcsTokenManagementService` as suggested to me by @b-fein, an interface for creating and renewing access tokens is provided independent of the used VCS, enabling extensibility for other VCSs in the future if necessary (see #8103).

While reworking the code for managing personal access tokens, I also fixed a todo stating that the programmatic GitLab API should be used instead of manual HTTP requests for creating personal access tokens. Unfortunately, listing and revoking personal access tokens is not (yet?) supported by this programmatic GitLab API, so I had to perform those requests manually.

Furthermore, note that although there exists a request type `Rotate` for effectively revoking and creating a personal access token in one single request instead of two (revoke+create), this was not used due to GitLab's use of so called `Automatic reuse detection` (see also https://docs.gitlab.com/ee/api/personal_access_tokens.html#automatic-reuse-detection): When rotating a personal access token, GitLab stores all previous access tokens and revokes the currently active one if there occurs a usage attempt of a prior access token. This feature is said to strengthen security against access token leaks, but it is unsuitable for our use case, where an accidental reuse of a prior access token is quite likely, e.g. when they are used to push/pull over HTTPS.

### Steps for Testing

Prerequisites:
- GitLab is correctly set up as VCS (also set `version-control-access-token` to `true` in the Artemis configuration)
- To make the manual testing practical, temporarily change the following timing constant in the code before running the server: In `VcsTokenRenewalService.java`, change line 58 from
    ```java
    @Scheduled(cron = "0  0  4 * * SUN") // Every sunday at 4 am
    ```
    to
    ```java
    @Scheduled(cron = "0 * * * * *") // Every minute
    ```
    such that the scheduled access token renewal job is run every minute.
- A connection to the Artemis database via a database client, for example using `psql` for a Postgres database. This is necessary to simulate an expiring access token.
- 1 Student with a personal access token. Let `<username>` be the login name of that student.

1. Start Artemis with the code changes from above.
2. Connect to the database.
3. Run
    ```sql
    SELECT login, vcs_access_token, vcs_access_token_expiry_date FROM jhi_user WHERE login = '<username>';
    ```
    to see the personal access token of the student and the expiry date.
4. Run
    ```sql
    UPDATE jhi_user SET vcs_access_token_expiry_date = '2024-03-20 00:00:00' WHERE login = '<username>';
    ```
    This simulates that the access token has expired or is about to expire (depending on the day of this test).
5. Wait two minutes.
6. Run
    ```sql
    SELECT login, vcs_access_token, vcs_access_token_expiry_date FROM jhi_user WHERE login = '<username>';
    ```
    and check that the personal access token has been updated and that the expiry date as been set to 365 days from running this test.

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| `GitLabPersonalAccessTokenManagementService` | 93%  | ✅  |
| `GitlabUserManagementService` | 94% |  ✅  |
| `VcsTokenRenewalService` | 87% (only two catch+log blocks not covered) |  ✅  |

## Summary by CodeRabbit

- **New Features**
    - Added two new methods for user repository to manage access token expiration and null access tokens.
    - Introduced `VcsTokenRenewalService` for renewing and creating VCS access tokens on a scheduled task.
- **Enhancements**
    - Improved GitLab user management by integrating token management services.
- **Tests**
    - Added tests for new GitLab personal access token management and VCS token renewal services.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabPersonalAccessTokenManagementService.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    private void renewVersionControlAccessToken(org.gitlab4j.api.models.User gitlabUser, User user, Duration newLifetime) {
+        GitLabPersonalAccessTokenListResponseDTO response = fetchPersonalAccessTokenId(gitlabUser.getId());
+
+        revokePersonalAccessToken(response.getId());
+
+        // Set access token to null for local user object, otherwise createAccessToken does nothing.
+        user.setVcsAccessToken(null);
+        user.setVcsAccessTokenExpiryDate(null);
+
+        createAccessToken(user, newLifetime);
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabPersonalAccessTokenManagementService.java:136]: This call then fetches the `gitlabUser` again from the GitLab API even though we have it already.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabPersonalAccessTokenManagementService.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    private void renewVersionControlAccessToken(org.gitlab4j.api.models.User gitlabUser, User user, Duration newLifetime) {
+        GitLabPersonalAccessTokenListResponseDTO response = fetchPersonalAccessTokenId(gitlabUser.getId());
+
+        revokePersonalAccessToken(response.getId());
+
+        // Set access token to null for local user object, otherwise createAccessToken does nothing.
+        user.setVcsAccessToken(null);
+        user.setVcsAccessTokenExpiryDate(null);
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/gitlab/GitLabPersonalAccessTokenManagementService.java:134]: Would it make sense to encapsulate this into a
```java
private void revokePersonalAccessToken(org.gitlab4j.api.models.User gitlabUser, User user) {…}
```
?

That new method would then be reusable for the planned future extensions and ensure the `user` object stays consistent with the token state.

File: src/test/java/de/tum/in/www1/artemis/service/connectors/vcs/VcsTokenRenewalServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            if (data.initialLifetimeDays() != null) {
+                user.setVcsAccessTokenExpiryDate(ZonedDateTime.now().plusDays(data.initialLifetimeDays()));
+            }
+            else {
+                user.setVcsAccessTokenExpiryDate(null);
+            }
+            userRepository.save(user);
+
+            gitlabUsers.add(new org.gitlab4j.api.models.User().withId(user.getId()).withUsername(user.getLogin()));
+        }
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/service/connectors/vcs/VcsTokenRenewalServiceTest.java:119]: This test method is quite long. Can you extract some parts into helper methods? Currently it is hard to see where the setup ends, where the actual method under test is called, and then where the assertions happen (up to line 150, line 151, and lines 152 onward).

E.g. the marked part into a `setupGitLabUsers(users)`.

**Comment by R3dError** [comment]
@b-fein Thank you for your review, I added your changes as requested.

**Review by b-fein — CHANGES_REQUESTED** [review]
Thanks for implementing the suggested changes. Code looks good now. :+1:

I noticed another small but important thing on the re-review though.

**Review by b-fein — APPROVED** [review]
Code looks good now. :+1:

**Review by chrisknedl — APPROVED** [review]
Tested on TS6. Works as expected, great work!

**Review by b-fein — APPROVED** [review]
We also tested this PR on a Artemis+GitLab test server at Uni Passau since testing this on a GitHub-deployed testserver is not possible.

**Review by krusche — APPROVED** [review]
Code looks good to me 👍

**Pull request merged**